modules/users/server/controllers/users/users.profile.server.controller.js

```diff
@@ -60,39 +60,83 @@ exports.changeProfilePicture = function (req, res) {
   var user = req.user;
   var upload = multer(config.uploads.profileUpload).single('newProfilePicture');
   var profileUploadFileFilter = require(path.resolve('./config/lib/multer')).profileUploadFileFilter;
+  var tmpImageUrl;
 
   // Filtering to upload only images
   upload.fileFilter = profileUploadFileFilter;
-
+  
   if (user) {
-    upload(req, res, function (uploadError) {
-      if (uploadError) {
-        return res.status(400).send({
-          message: 'Error occurred while uploading profile picture'
-        });
-      } else {
-        user.profileImageURL = config.uploads.profileUpload.dest + req.file.filename;
+    tmpImageUrl = user.profileImageURL;
+    uploadImage()
+      .then(updateUser)
+      .then(deleteOldImage)
+      .then(login)
+      .then(function () {
+        res.json(user);
+      })
+      .catch(function (err) {
+        res.status(400).send(err);
+      });
+  } else {
+    res.status(400).send({
+      message: 'User is not signed in'
+    });
+  }
+
+  function uploadImage () {
+    return new Promise(function (resolve, reject) {
+      upload(req, res, function (uploadError) {
+        if (uploadError) {
+          reject({
+            message: 'Error occurred while uploading profile picture'
+          });
+        } else {
+          resolve();
+        }
+      });
+    });
+  }
+
+  function updateUser () {
+    return new Promise(function (resolve, reject) {
+      user.profileImageURL = config.uploads.profileUpload.dest + req.file.filename;
+      user.save(function (err, theuser) {
+        if (err) {
+          reject(err);
+        } else {
+          resolve();
+        }
+      });
+    });
+  }
 
-        user.save(function (saveError) {
-          if (saveError) {
-            return res.status(400).send({
-              message: errorHandler.getErrorMessage(saveError)
+  function deleteOldImage () {
+    return new Promise(function (resolve, reject) {
+      if (tmpImageUrl !== User.schema.path('profileImageURL').defaultValue) {
+        fs.unlink(tmpImageUrl, function (unlinkError) {
+          if (unlinkError) {
+            reject({
+              message: 'Error occurred while uploading profile picture'
```

Comment:
Also, what info does `unlinkError` contain? Would it be useful to provide that with the reject?